# Part 1. Equation of a Slime

How many late days are you using for this assignment?
0

In [87]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split , cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

## 1. Loading the dataset

In [88]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
science_df = pd.read_csv("science_data_large.csv")
print(science_df.head(15))
print(science_df.describe())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
       Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500

## 2. Splitting the dataset

In [89]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = science_df[['Mols KCL', 'Temperature °C']]  # Features
y = science_df['Size nm^3']              # Label

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [90]:
# Use sklearn to train a model on the training set
reg_model = LinearRegression()
reg_model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = pd.DataFrame({'Mols KCL': [500], 'Temperature °C': [500]})
prediction = reg_model.predict(sample)
print("Prediction:", prediction)
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
model_score = reg_model.score(X_test, y_test)
print("Model score:", model_score)
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coeff = reg_model.coef_
intercept = reg_model.intercept_

print("Coeffecients :" ,coeff,"Intercepts :", intercept )

Prediction: [540029.26034545]
Model score: 0.8552472077276096
Coeffecients : [1032.69506649  866.14641337] Intercepts : -409391.47958340833


Write the linear equation of a slime: (example equation: $E = mc^2$) 
$$ \text{Size nm}^3 = -409391.48 + 1032.70 \cdot \text{(Mols KCL)} + 866.15 \cdot \text{(Temperature  °C)} $$


Report on score and explain meaning: The score metric is R^2, where the range varies from 0-1 with higher scores representing better precdictions in the linear relationship. The model has a score of .855, meaning that the model is doing well in understanding and capturing the underlying relationship.  

## 4. Use Cross Validation

In [91]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
scores = cross_val_score(reg_model, X, y, cv=5)
print("Cross val scores:", scores)
# Report on their finding and their significance
coeff = reg_model.coef_
intercept = reg_model.intercept_

print("Coeffecients :" ,coeff,"Intercepts :", intercept )

Cross val scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Coeffecients : [1032.69506649  866.14641337] Intercepts : -409391.47958340833


Write findings here: These scores the R^2 values over 5 different splits of the dataset. Since the range is high (.83-.87), we can say that the model is able to properly understand and interpret the relationship, as said above

## 5. Using Polynomial Regression

In [92]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
from sklearn.preprocessing import PolynomialFeatures
poly = PolynomialFeatures(degree=2)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

model_poly = LinearRegression()
model_poly.fit(X_train_poly, y_train)

scores_poly = cross_val_score(model_poly, X_train_poly, y_train, cv=5)
print("Polynomial scores :", scores_poly)
# Report on the metrics and output the resultant equation as you did in Part 3.
poly_coeff = model_poly.coef_
poly_intercept = model_poly.intercept_

print("Coeffecients :" ,poly_coeff,"Intercepts :", poly_intercept )

Polynomial scores : [1. 1. 1. 1. 1.]
Coeffecients : [ 0.00000000e+00 -1.27197610e-07  1.20000000e+01  2.85714287e-02
  2.00000000e+00  1.26449962e-11] Intercepts : 2.0478502847254276e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)
$$\text{Size nm}^3 = 2.0478502847254276 \times 10^{-5} - 1.27197610 \times 10^{-7} \cdot \text{(Temperature)} + 12.0 \cdot \text{(Temperature)}^2 + 0.0285714287 \cdot \text{(Mols KCL)}^2 + 2.0 \cdot \text{(Mols KCL)} \cdot \text{(Temperature)}$$

Report on the score and interpret:
THe score I got is a series of 1.0s, meaning that my model has achieved a perfect or near-perfect fit to the dataset. I believe this is a sign of overfitting, the model is perfectly trained to the samle data but if I were to introduce new data it would be inaccurate. 

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [93]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import roc_curve, roc_auc_score
import matplotlib.pyplot as plt
# Load the dataset. Then train and evaluate the classification models.


ckd_df = pd.read_csv("ckd_feature_subset.csv")
X = ckd_df[['age', 'bp', 'wbcc', 'appet_poor', 'appet_good', 'rbcc']]
y = ckd_df['Target_ckd']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.1, random_state=42)

models = {
    'Logistic Regression': LogisticRegression(max_iter=1000),
    'Support Vector Machines': SVC(max_iter=1000),  
    'k-Nearest Neighbors': KNeighborsClassifier(),
    'Neural Network': MLPClassifier(max_iter=1000)
}

cv_results = pd.DataFrame()

# Perform 5-fold cross-validation for each model and collect accuracy scores
for model_name, model in models.items():
    scores = cross_val_score(model, X, y, cv=5, scoring='accuracy')
    cv_results[model_name] = scores  # Add the scores to the DataFrame

# Display the results
print("Cross-Validation Accuracy Scores:")
print(cv_results)

#Holistic score result 
mean_scores = cv_results.mean().rename('Mean Accuracy')
std_scores = cv_results.std().rename('Standard Deviation')
summary = pd.concat([mean_scores, std_scores], axis=1)
print("\nSummary of Cross-Validation Results:")
print(summary)


Cross-Validation Accuracy Scores:
   Logistic Regression  Support Vector Machines  k-Nearest Neighbors  \
0             0.870968                 0.903226             0.903226   
1             0.838710                 1.000000             0.967742   
2             0.870968                 0.870968             0.903226   
3             0.933333                 1.000000             0.966667   
4             0.766667                 0.866667             0.900000   

   Neural Network  
0        0.967742  
1        0.935484  
2        0.903226  
3        0.966667  
4        0.933333  

Summary of Cross-Validation Results:
                         Mean Accuracy  Standard Deviation
Logistic Regression           0.856129            0.060638
Support Vector Machines       0.928172            0.067075
k-Nearest Neighbors           0.928172            0.035658
Neural Network                0.941290            0.026877


## Results and Conclusion for Classification Experiments

1. Logistic Regression - The accuracy for logistic regression across the 5 folds ranged from .7667 -.9333. The model had a wide range when it came to accuracy, showing that even though it can distinguish between classes, it may struggle within several subdatasets. 
2. SVM - The SVM showed strong and consistent performance within the 5 folds as it ranged from from .8709678 - 1.000. 2 of the folds had 1.0000, which show that there is a chance the model is being overfit on the data, that it is not able to build proper patterns and is following hte noise in the model.
3. K-NN - KNN also had strong and consistent scores within the 5 folds as it ranged from .90 - .967742. The model was thus able to properly distinguise between the classes and does not show unusual signs of being overfit, making it a strong contender for best model in this list. 
4. Neural Network ( MLP CLassifier) - The Neural Network had the best scores as it ranged from .9032 - .9677, showing lower variability than KNN while also not being overfit to the given data. These scores show that the model is capable of generalizing and it is able to keep this high accuracy without much variance. 

In [94]:
# Experiments with Neural Network.
# Configuration 1: Simple network
nn_configs ={
'Config 1':MLPClassifier(hidden_layer_sizes=(10,), max_iter=1000, random_state=42),

# Configuration 2: Deeper network
'Config 2' :MLPClassifier(hidden_layer_sizes=(50, 25), max_iter=1000, random_state=42),
# Configuration 3: Single layer network with tanh activation
'Config 3' : MLPClassifier(hidden_layer_sizes=(100,), activation='tanh', max_iter=1000, solver='adam', early_stopping=True, random_state=42)
}

nn_results = pd.DataFrame()

for config_name, model in nn_configs.items():
    scores = cross_val_score(model, X, y, cv=5, scoring='accuracy')
    nn_results[config_name] = scores  


print("Neural Network Configuration Accuracy Scores:")
print(nn_results)



c:\Users\manov\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\manov\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\manov\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\manov\anaconda3\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\manov\anaconda3\Lib\site-packages\sklearn\neural_network\_m

Neural Network Configuration Accuracy Scores:
   Config 1  Config 2  Config 3
0  0.903226  1.000000  0.677419
1  0.967742  0.967742  0.838710
2  0.870968  0.935484  0.870968
3  0.933333  1.000000  0.933333
4  0.800000  0.933333  0.733333


## Results and Conclusion for Neural Network Experiments

The best performaning configuration was config 2 which has the highest average highest accuracy of about 96.73% with low variability, showing that a deep network with 2 layers might yield better results in classifying the data. Config 1 was similar but did not yield the best results at an average of 91.42%. Config 3 had the lowest average of 83% , which signals that the tanh activatiohn might not help with this dataset.   